src/openMVG/robust_estimation/rand_sampling.hpp

```diff
@@ -25,85 +25,156 @@
-* \param total_samples The number of samples available.
-* \param samples       num_samples of numbers in [0, total_samples) is placed
-*                      here on return.
-* \warning Argument values should respect: num_samples <= total_samples
-*/
+ * @brief Generate a unique random samples without replacement in the 
+ * range [lowerBound upperBound).
+ * It is modeled after Matlab function with the same name, and it tries to optimize
+ * the generation of the random samples: if the number of required samples is a
+ * large ratio of the range, then it shuffles a vector containing all the numbers 
+ * in the range and it takes the first numSamples elements. Otherwise it proceeds
+ * by drawing random numbers until the numSamples elements are generated, using  
+ * Robert Floyd's algorithm.
+ * 
+ * @param[in] lowerBound The lower bound of the range.
+ * @param[in] upperBound The upper bound of the range (not included).
+ * @param[in] numSamples Number of unique samples to draw.
+ * @return samples The vector containing the samples.
+ */
 template<typename IntT>
-static void UniformSample(
-  size_t num_samples,
-  size_t total_samples,
-  std::set<IntT> *samples)
+inline std::vector<IntT> randSample(IntT lowerBound,
+                                    IntT upperBound,
+                                    IntT numSamples)
 {
-  assert(num_samples <= total_samples);
+  const auto rangeSize = upperBound - lowerBound;
+  
+  assert(lowerBound < upperBound);
+  assert(numSamples <= rangeSize);
+  static_assert(std::is_integral<IntT>::value, "Only integer types are supported");
+
+  
   std::random_device rd;
-  std::default_random_engine e1(rd());
-  std::uniform_int_distribution<IntT> uniform_dist(0, total_samples-1);
-  while (samples->size() < num_samples)
+  std::mt19937 generator(rd());
+
+  if(numSamples * 1.5 > rangeSize)
+  {
+    // if the number of required samples is a large fraction of the range size
+    // generate a vector with all the elements in the range, shuffle it and 
+    // return the first numSample elements.
+    // this should be more time efficient than drawing at each time.
+    std::vector<IntT> result(rangeSize);
+    std::iota(result.begin(), result.end(), lowerBound);
+    std::shuffle(result.begin(), result.end(), generator);
+    result.resize(numSamples);
+    return result;
+  }
+  else
   {
-    IntT sample = uniform_dist(e1);
-    samples->insert(sample);
+    // otherwise if the number of required samples is small wrt the range
+    // use the optimized Robert Floyd algorithm.
+    // this has linear complexity and minimize the memory usage.
+    std::unordered_set<IntT> samples;
+    for(IntT d = upperBound - numSamples; d < upperBound; ++d)
+    {
+      IntT t = std::uniform_int_distribution<>(0, d)(generator) + lowerBound;
+      if(samples.find(t) == samples.end())
+        samples.insert(t);
+      else
+        samples.insert(d);
+    }
+    assert(samples.size() == numSamples);
```

Comment:
but actually a good point is that i can use directly a `std::vector` instead of a `std::unordered_set` and spare the final copy.